sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,325 @@
+                spis: Default::default(),
+                uarts: Default::default(),
+            })
+        };
+        Ok(result)
+    }
+
+    /// Locates the /dev/ttyUSBn node corresponding to a given interface in the sys directory
+    /// tree, e.g. /sys/bus/usb/devices/1-4/1-4:1.0 .
+    fn find_tty(path: &Path) -> Result<PathBuf> {
+        for entry in fs::read_dir(path)? {
+            let dir_entry = entry?;
+            match dir_entry.file_name().into_string() {
+                Ok(filename) => {
+                    if filename.starts_with("tty") {
+                        return Ok(PathBuf::from("/dev").join(dir_entry.file_name()));
+                    }
+                }
+                _ => ()
+            }
+        }
+        Err(Error::CommunicationError.into())
+    }
+}
+
+impl Inner {
+
+    /// Send a command to HyperDebug firmware, with a callback to receive any output.
+    pub fn execute_command(
+        &self,
+        cmd: &str,
+        mut callback: impl FnMut(&str)) -> Result<()>
+    {
+        use std::io::Read;
+        use std::io::Write;
+        let mut port = serialport::new(self.console_tty.to_str().ok_or(Error::UnicodePathError)?, 115_200)
+            .timeout(std::time::Duration::from_millis(100))
+            .open().expect("Failed to open port");
+
+        // Ideally, we would invoke Linux flock() on the serial
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf: [u8; 128] = [0; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    print!("Discarded {} characters: {:?}\n", rc, &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            break;
+                        }
+                        _ => {
+                            print!("Error reading: {:?}\n", error);
+                            return Err(error.into())
+                        }
+                    }
+                }
+            }
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
+        let mut seen_echo = false;
+        let mut len: usize = 0;
+        loop {
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
+                        for i in 0..len {
+                            if buf[i] == 10 {
+                                // Newline
+                                let mut end = i;
+                                if end > 0 && buf[end - 1] == 13 {
```

Comment:
Is this protocol fundamentally line-based?  If so, would `BufReader` which has methods like `read_line`, `read_to_string` and `read_until` make this less complex?

Reply:
It is line based, but I need to detect when a prompt was printed, and not terminated with a newline, which I suppose means that it would not help to use BufReader.

I have simplified the logic slightly, by removing some logic that was added in response to initial characters sometimes being lost (I later realized that the loss was caused by a minicom running in the background on the same ttyUSBx device, snubbing some characters now and then.)